jobserver/api/releases.py

```diff
@@ -556,3 +558,73 @@ class WorkspaceStatusAPI(RetrieveAPIView):
 
     class serializer_class(serializers.Serializer):
         uses_new_release_flow = serializers.BooleanField()
+
+
+class TokenAuthenticationAPI(APIView):
+    authentication_classes = []
+
+    class serializer_class(serializers.Serializer):
+        user = serializers.CharField()
+        token = serializers.CharField()
+
+    class Level4AuthenticatedUser(serializers.Serializer):
+        username = serializers.CharField()
+        fullname = serializers.CharField()
+        workspaces = serializers.ListField(
+            child=serializers.CharField(),
+            default=[],
+        )
+        output_checker = serializers.BooleanField()
+        staff = serializers.BooleanField()
+
+    def post(self, request):
+        if getattr(request, "backend", None) is None:
+            logger.info("API Login with token failed because request not from backend")
+            raise NotAuthenticated()
+
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        data = serializer.data
+
+        try:
+            # user can be username or email
+            user = users.validate_login_token(data["user"], data["token"])
+        except (
+            User.DoesNotExist,
+            users.InvalidTokenUser,
+            users.BadLoginToken,
+            users.ExpiredLoginToken,
```

Comment:
Minor aside, and I know this is just replicating the existing token auth code, but I found it surprising there wasn't a base exception class we could catch here. It seems odd the caller needs to enumerate the specific ways the login could fail.

Reply:
Fair point, will add